### PR TITLE
Inference device selector

### PR DESCRIPTION
**Pull request by camiloHimura**
#5043

### Summary

https://github.com/user-attachments/assets/e714f9cf-fd70-4bf8-8c3b-537750bbf8e3

### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import { $api } from '../../../api/client';
+import { usePipeline } from '../../../hooks/api/pipeline.hook';
+import { useProjectIdentifier } from '../../../hooks/use-project-identifier.hook';
+
+export const InferenceDevices = () => {
+    const { data: devices } = $api.useSuspenseQuery('get', '/api/system/devices/inference');
+    const { data: pipeline } = usePipeline();
+    const projectId = useProjectIdentifier();
+    const [selectedKey, setSelectedKey] = useState<Key | null>(pipeline.device.toLowerCase() ?? null);
```
[the remark this answers is not in the document]
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx]: updated

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    const handleChange = (key: Key | null) => {
+        if (key === null) {
+            return;
+        }
+
+        setSelectedKey(key);
+        updatePipeline.mutate({
+            params: { path: { project_id: projectId } },
+            body: { device: key },
```
[the remark this answers is not in the document]
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx]: updated

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        });
+    };
+
+    return (
+        <Picker
+            maxWidth={'size-3000'}
```
[the remark this answers is not in the document]
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx]: updated

File: application/ui/src/features/inference/header/active-model.component.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                }}
+            >
+                Model:
+            </Text>
+            <Text
+                UNSAFE_style={{
+                    color: 'var(--spectrum-global-color-gray-700)',
+                }}
+            >
+                {!isEmpty(modelsQuery.data) ? modelsQuery.data[0].architecture : 'Unknown'}
```
Comment (jpggvilaca) [inline application/ui/src/features/inference/header/active-model.component.tsx]: the first one might not be the active one. Let's use the hook that we have to get the active model instead

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import { usePipeline } from '../../../hooks/api/pipeline.hook';
+import { useProjectIdentifier } from '../../../hooks/use-project-identifier.hook';
+
+export const InferenceDevices = () => {
+    const { data: devices } = $api.useSuspenseQuery('get', '/api/system/devices/inference');
+    const { data: pipeline } = usePipeline();
+    const projectId = useProjectIdentifier();
+    const [selectedKey, setSelectedKey] = useState<Key | null>(pipeline.device);
+
+    const updatePipeline = $api.useMutation('patch', '/api/projects/{project_id}/pipeline', {
```
Comment (jpggvilaca) [inline application/ui/src/features/inference/header/inference-devices.component.tsx]: np we can use `usePatchPipeline` from pipeline hooks

**Review by jpggvilaca — DISMISSED** [review]

File: application/ui/src/features/inference/header/active-model.component.tsx (hunk 4 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/active-model.component.tsx]: updated

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 5 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx]: updated

File: application/ui/src/features/inference/header/active-model.component.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+                }}
+            >
+                Model:
+            </Text>
+            <Text
+                UNSAFE_style={{
+                    color: 'var(--spectrum-global-color-gray-700)',
+                }}
+            >
+                {isEmpty(activeModelArchitectureId) ? 'Unknown' : activeModelArchitectureId}
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/active-model.component.tsx:26]: When is it possible to have unknown model?

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import { useState } from 'react';
+
+import { Item, Key, Picker, toast } from '@geti/ui';
+
+import { $api } from '../../../api/client';
+import { usePatchPipeline, usePipeline } from '../../../hooks/api/pipeline.hook';
+import { useProjectIdentifier } from '../../../hooks/use-project-identifier.hook';
+
+export const InferenceDevices = () => {
+    const { data: devices } = $api.useSuspenseQuery('get', '/api/system/devices/inference');
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.component.tsx:13]: Could we extract api hooks to `api` dir? So we don't mix responsibilities of api hooks with component

File: application/ui/src/features/inference/header/inference-devices.test.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            http.get('/api/system/devices/inference', () => HttpResponse.json(devicesResponse)),
+            http.get('/api/projects/{project_id}/pipeline', () => HttpResponse.json(pipelineResponse)),
+            http.patch('/api/projects/{project_id}/pipeline', () => {
+                pipelinePatchSpy();
+                return HttpResponse.json({}, { status });
+            })
+        );
+
+        render(
+            <TestProviders client={new QueryClient()}>
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.test.tsx:40]: Why new client needs to be passed?

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 7 above)
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.component.tsx:13]: We use suspense query but we don't have a suspense boundary around this component which will remount this and probably more components.

File: application/ui/src/features/inference/header/inference-devices.component.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    const { data: pipeline } = usePipeline();
+    const projectId = useProjectIdentifier();
+    const [selectedKey, setSelectedKey] = useState<Key | null>(pipeline.device);
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.component.tsx:16]: Is it possible that `pipeline.device` is not part of `devices`? (it might be the edge case but it'd be better to double check)

File: application/ui/src/features/inference/header/inference-devices.test.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import userEvent from '@testing-library/user-event';
+import { getMockedPipeline } from 'mocks/mock-pipeline';
+import { HttpResponse } from 'msw';
+import { TestProviders } from 'test-utils/render';
+
+import { http } from '../../../api/utils';
+import { server } from '../../../msw-node-setup';
+import { InferenceDevices } from './inference-devices.component';
+
+vi.mock('hooks/use-project-identifier.hook', () => ({ useProjectIdentifier: () => '123' }));
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.test.tsx:15]: For later: we can use test renderer from prompt where we specify the url (together with project id) so we would not need to mock `useProjectIdentifier`.

File: application/ui/src/features/inference/header/inference-devices.test.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    it('updates device on selection change', async () => {
+        const pipelinePatchSpy = renderApp();
+
+        await screen.findByLabelText('inference compute');
+
+        await userEvent.click(screen.getByRole('button', { name: /inference compute/i }));
+        await userEvent.click(screen.getByRole('option', { name: /GPU/i }));
+
+        await waitFor(() => {
+            expect(pipelinePatchSpy).toHaveBeenCalled();
```
Comment (dwesolow) [inline application/ui/src/features/inference/header/inference-devices.test.tsx:67]: Instead of checking if the spy is called it'd be more appropriate to check what is the boy of the `patch` request.

File: application/ui/src/features/inference/header/active-model.component.tsx (hunk 6 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/active-model.component.tsx:26]: I’m not sure, I just moved the component to its own file, and I don’t know why it was implemented that way originally.

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 7 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx:13]: why?

File: application/ui/src/features/inference/header/inference-devices.test.tsx (hunk 8 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.test.tsx:40]: It was reusing the same query configuration and returning responses from the preview test configurations.

File: application/ui/src/features/inference/header/inference-devices.test.tsx (hunk 10 above)
Comment (jpggvilaca) [reply application/ui/src/features/inference/header/inference-devices.test.tsx:15]: Context: https://github.com/open-edge-platform/geti-prompt/blob/main/application/ui/src/test-utils/render.tsx

**Review by jpggvilaca — APPROVED** [review]

File: application/ui/src/features/inference/header/inference-devices.test.tsx (hunk 11 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.test.tsx:67]: Perhaps, but not long ago I spent some time trying to make that assertion work, and it turned out to be quite problematic. Because of that, I went with the easiest approach.

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 9 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx:16]: No, that’s not possible. And even if it were, I believe this would be a backend issue rather than something on our side.

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 7 above)
Comment (camiloHimura) [reply application/ui/src/features/inference/header/inference-devices.component.tsx:13]: I'll handle it on my next PR

File: application/ui/src/features/inference/header/inference-devices.test.tsx (hunk 11 above)
Comment (dwesolow) [reply application/ui/src/features/inference/header/inference-devices.test.tsx:67]: I did that approach a lot in prompt and I haven't noticed any issue. What was the issue that you got?

File: application/ui/src/features/inference/header/inference-devices.component.tsx (hunk 7 above)
Comment (dwesolow) [reply application/ui/src/features/inference/header/inference-devices.component.tsx:13]: To not mix responsibilities.

File: application/ui/src/features/inference/header/inference-devices.test.tsx (hunk 8 above)
Comment (dwesolow) [reply application/ui/src/features/inference/header/inference-devices.test.tsx:40]: If that's the case we should not fix the problem locally, but for each case (always pass new query client inside the test providers)
